log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2016-2021 Talsma ICT
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package nl.talsmasoftware.context.log4j2;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.logging.log4j.ThreadContext;
+
+/**
+ * Represents a snapshot of the data of the Log4j 2 {@link ThreadContext} of a
+ * specific thread at a certain point in the past.
+ */
+public class ThreadContextData {
+    private final Map<String, String> contextMap;
+    private final List<String> contextStack;
+
+    /**
+     * Applies the data to the current thread.
+     *
+     * @param data data to apply, may be {@code null}
+     * @param overwrite whether all existing data should overwritten
+     */
+    static void applyToCurrentThread(ThreadContextData data, boolean overwrite) {
+        if (overwrite) {
+            ThreadContext.clearAll();
+        }
+        if (data != null) {
+            data.applyToCurrentThread();
+        }
+    }
+
+    /**
+     * Applies the data on top of the {@code ThreadContext} of the current thread.
+     * Conflicting entries are overwritten but all other existing data is kept.
+     */
+    void applyToCurrentThread() {
+        ThreadContext.putAll(contextMap);
+
+        // There is currently no method for pushing a collection, therefore have to
+        // push one by one
+        for (String element : contextStack) {
+            ThreadContext.push(element);
+        }
+    }
+
+    private ThreadContextData(Map<String, String> contextMap, List<String> contextStack) {
+        this.contextMap = Collections.unmodifiableMap(contextMap);
```

Comment:
The methods `getContextMap()` and `getContextStack()` publicly expose the values of these fields and the user should not be able to manipulate them; not even the `ThreadContextData` class itself should be able to manipulate them, since the data represent an unmodifiable snapshot.

It sould also be possible to wrap them in unmodifiable collections every time `getContextMap()` and `getContextStack()` is called, but that would only create unnecessary overhead.